src/main/java/org/reaktivity/specification/socks/internal/SocksFunctions.java

```diff
@@ -126,6 +225,62 @@ public String getPrefixName()
         }
     }
 
+    public static void fillInBytes(byte[] addressBytes, int index, String ipv6Group)
+    {
+        int ipv6GroupLength = ipv6Group.length();
+        if  (ipv6GroupLength < 3)
+        {
+            addressBytes[2 * index] = 0;
+            addressBytes[2 * index + 1] = (byte) Integer.parseInt(ipv6Group, 16);
+        }
+        else
+        {
+            int mid = ipv6GroupLength/2;
+            addressBytes[2 * index] = (byte) Integer.parseInt(ipv6Group.substring(0, mid), 16);
+            addressBytes[2 * index + 1] =
+                (byte) Integer.parseInt(ipv6Group.substring(mid, ipv6GroupLength), 16);
+        }
+    }
```

Comment:
Use `Byte.parseByte` instead of `Integer.parseInt` to avoid the cast to `byte`.
Avoid the special case for `ipv6GroupLength < 3`.
Consider one looping algorithm that can handle variable length groups, converting them from hex to bytes.
`index` would advance by `2` for each `byte` converted to `hex`
this will eliminate to keep recalculating via `/` and `*`, instead streaming through the group characters

Reply:
for `byte.parseByte`, the range will from 128 to -127, which doesn't cover all hex range, like `ff`, should I keep `integer.parseint` or do something to `string` or `character`?

Reply:
Look at the implementation of `Byte.parseByte` and compare it to your earlier approach of `Integer.parseInt` cast to `byte`.

Reply:
`Byte.pareseByte` has almost the same structure as mine, but I think this line will cost error:
`if (i < MIN_VALUE || i > MAX_VALUE)
            throw new NumberFormatException(
                "Value out of range. Value:\"" + s + "\" Radix:" + radix);`
For example, `Byte.parseByte("ff", 16)`, will  call `Integer.parseInt("ff", 16)` first, the result will be 255, so `i = 255`. Then `255 > 127` which will throw an exception.

Reply:
Agreed, and we don't want to trigger that exception, so you should create a static `parseByte` method that takes the same args, but replaces their min-value and max-value checks with a length check on the input string to ensure that `0 < length <= 2` instead.

Reply:
I also use `parseInteger` for `IPV4`, can I change the range that `0 < length <= 3`, so the method will able to handle all `ipv4` and `ipv6` cases. For `ipv6` I will ensure length is 2.

Reply:
I think you want a separate `parseShort` for that.

Reply:
done